### PR TITLE
fix(Partials): correctly document properties for partialized structures

### PR DESCRIPTION
**Pull request by NotSugden**
**Please describe the changes this PR makes and why it should be merged:**
List of changes in this PR:
- document properties that can't be null on `Partialize`d structures
  - some properties defined in each Partial* are overwritten to specify what they should be if the structure is Partial
  - `client`, `createdAt`, `createdTimestamp` as non-nullable
    - the `client` property exists on everything that is `Partialize`d, and isn't nullable
    - `createdAt` and `createdTimestamp` are deconstructed from the ID, which isn't nullable, so these properties can't be null
- correctly documents that the functions can't be null
- `Message#channel` can be a `NewsChannel`, and the typings were updated to reflect that
- `GuildMember#roles` is readonly, and the typings were updated to reflect that

- PartialTextBasedChannelFields#(lastMessageChannelID, lastPinAt, lastPinTimestamp):
  - lastMessageChannelID was moved to the class definitions for User and GuildMember because it only existed on those classes
  - lastPin(At, Timestamp) was moved to TextBasedChannelFields as they didn't exist in classes that implemented PartialTextBasedChannelFields
- User#tag was updated to return null if the user is partial

- Extra information:
  -  it doesn't seem that a PartialDMChannel can be emitted into `channelCreate`, which is why it has been seperated from `channelDelete` in the typings
  - `PartialChannel` was renamed to `PartialDMChannel` as at the moment only DM channels can be partial

**Status**

- [x] Code changes have been tested against the Discord API, or there are no code changes
- [x] I know how to update typings and have done so, or typings don't need updating

**Semantic versioning classification:**

- [ ] This PR changes the library's interface (methods or parameters added)
  - [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

File: src/structures/User.js [hunk 1]
```diff
@@ -170,7 +170,7 @@ class User extends Base {
    * @readonly
    */
   get tag() {
```
Comment (SpaceEEC) [inline src/structures/User.js]: Shouldn't this be nullable then too?

File: typings/index.d.ts [hunk 2]
```diff
@@ -913,7 +915,7 @@ declare module 'discord.js' {
     public setName(name: string, reason?: string): Promise<GuildEmoji>;
   }
 
-  export class GuildMember extends PartialTextBasedChannel(Base) {
+  export class GuildMember extends (PartialTextBasedChannel(Base) & { lastMessageChannelID: Snowflake | null }) {
```
Comment (SpaceEEC) [inline typings/index.d.ts]: Wouldn't adding `lastMessageChannelID` as regular field be better for readability?

File: typings/index.d.ts [hunk 3]
```diff
@@ -1569,7 +1571,7 @@ declare module 'discord.js' {
     public fetchWebhooks(): Promise<Collection<Snowflake, Webhook>>;
   }
 
-  export class User extends PartialTextBasedChannel(Base) {
+  export class User extends (PartialTextBasedChannel(Base) & { lastMessageChannelID: Snowflake | null }) {
```
Comment (SpaceEEC) [inline typings/index.d.ts]: Same here.

File: typings/index.d.ts (hunk 2 above)
Comment (NotSugden) [reply typings/index.d.ts]: oh, i meant to fix that but forgot, not sure why i did it like that in the first place

File: typings/index.d.ts [hunk 4]
```diff
@@ -926,7 +926,8 @@ declare module 'discord.js' {
     public readonly id: Snowflake;
     public readonly joinedAt: Date | null;
     public joinedTimestamp: number | null;
-    public readonly kickable: boolean;
+		public readonly kickable: boolean;
+		public lastMessageID: Snowflake | null;
```
Comment (SpaceEEC) [inline typings/index.d.ts]:
```suggestion
		public lastMessageChannelID: Snowflake | null;
```

File: typings/index.d.ts [hunk 5]
```diff
@@ -1592,7 +1583,8 @@ declare module 'discord.js' {
     public discriminator: string;
     public readonly defaultAvatarURL: string;
     public readonly dmChannel: DMChannel;
-    public id: Snowflake;
+		public id: Snowflake;
+		public lastMessageID: Snowflake | null;
```
Comment (SpaceEEC) [inline typings/index.d.ts]:
```suggestion
		public lastMessageChannelID: Snowflake | null;
```

File: typings/index.d.ts (hunk 4 above)
Comment (NotSugden) [reply typings/index.d.ts]: fixed

**Pull request merged**